### PR TITLE
[SDK-3137] Add ASPNET Owin Api Interactive QuickStart

### PR DESCRIPTION
**Pull request by frederikprijck**
Adds a new Interactive Quickstart for ASP.NET Owin Web Api

### CODE REVIEW

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+useCase: quickstart
+interactive: true
+files:
+  - files/startup
+  - files/openid-connect-signing-key-resolver
+  - files/scope-authorize-attribute
+  - files/api-controller
+---
+
+# Add Authorization to an ASP.NET Owin Web API application.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/webapi-owin/interactive.md]: Style guide: titles should be sentence case

```suggestion
# Add authorization to an ASP.NET Owin Web API application.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+Auth0 allows you to quickly add authentication and gain access to user profile information in your application. This guide demonstrates how to integrate Auth0 with any new or existing ASP.NET Owin Web API application using the `Microsoft.Owin.Security.Jwt` package.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/webapi-owin/interactive.md]: The title says "authorization" but here it starts "quickly add authentication" - which one is it?

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+<%= include('../../../_includes/_api_auth_intro') %>
+
+## Define permissions
+<%= include('../_includes/_api_scopes_access_resources') %>
+
+
+## Install dependencies
+
+To use Auth0 Access Tokens with ASP.NET Owin Web API you will use the OWIN JWT Middleware which is available in the `Microsoft.Owin.Security.Jwt` NuGet package.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/webapi-owin/interactive.md]: Style guide (word list): access tokens should be lowercase

```suggestion
To use Auth0 access tokens with ASP.NET Owin Web API you will use the OWIN JWT Middleware which is available in the `Microsoft.Owin.Security.Jwt` NuGet package.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+
+This class is then used as `TokenValidationParameters.IssuerSigningKeyResolver` while configuring the middleware in `Startup.cs`.
+
+:::note
+Such a custom resolver was previously published as part of the `Auth0.OpenIdConnectSigningKeyResolver` package through Nuget. As [this package is not available anymore](https://github.com/auth0/auth0-aspnet-owin/blob/master/SECURITY-NOTICE.md), you will need to provide this yourself.
+:::
+
+## Validate scopes {{{ data-action=code data-code="ScopeAuthorizeAttribute.cs" }}}
+
+The JWT middleware verifies that the Access Token included in the request is valid; however, it doesn't yet include any mechanism for checking that the token has the sufficient **scope** to access the requested resources.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/webapi-owin/interactive.md]: Style guide (word list): access token should be lowercase

```suggestion
The JWT middleware verifies that the access token included in the request is valid; however, it doesn't yet include any mechanism for checking that the token has the sufficient **scope** to access the requested resources.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+
+Create a class called `ScopeAuthorizeAttribute` which inherits from `System.Web.Http.AuthorizeAttribute`. This Authorization Attribute will check that the `scope` claim issued by your Auth0 tenant is present, and if so it will ensure that the `scope` claim contains the requested scope.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/webapi-owin/interactive.md]: Should Authorization be in backticks?

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+
+## Protect API endpoints {{{ data-action=code data-code="ApiController.cs" }}}
+
+<%= include('../_includes/_api_endpoints') %>
+
+The JWT middleware integrates with the standard ASP.NET Authentication and Authorization mechanisms, so you only need to decorate your controller action with the `[Authorize]` attribute to secure an endpoint.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
The JWT middleware integrates with the standard ASP.NET authentication and authorization mechanisms, so you only need to decorate your controller action with the `[Authorize]` attribute to secure an endpoint.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+To ensure that a scope is present in order to call a particular API endpoint, you simply need to decorate the action with the `ScopeAuthorize` attribute and pass the name of the required `scope` in the `scope` parameter.
+
+::::checkpoint
+
+:::checkpoint-default
+
+Now that you have configured your application, run your application to verify that:
+* `GET /api/public` is available for non-authenticated requests.
+* `GET /api/private` is available for authenticated requests.
+* `GET /api/private-scoped` is available for authenticated requests containing an Access Token with the `read:messages` scope.
```
Comment (stevehobbsdev) [inline articles/quickstart/backend/webapi-owin/interactive.md]: Style guide (word list): access token should be lowercase

```suggestion
* `GET /api/private-scoped` is available for authenticated requests containing an access token with the `read:messages` scope.
```

**Review by stevehobbsdev — CHANGES_REQUESTED** [review]
LGTM, just left a few style guide-related suggestions.

File: articles/quickstart/backend/webapi-owin/interactive.md (hunk 5 above)
Comment (frederikprijck) [reply articles/quickstart/backend/webapi-owin/interactive.md]: I removed it as I believe it was redundant.

**Review by stevehobbsdev — DISMISSED** [review]

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+interactive: true
+files:
+  - files/startup
+  - files/openid-connect-signing-key-resolver
+  - files/scope-authorize-attribute
+  - files/api-controller
+---
+
+# Add authorization to an ASP.NET Owin Web API application.
+Auth0 allows you to quickly add authorization to any kind of application. This guide demonstrates how to integrate Auth0 with any new or existing ASP.NET Owin Web API application using the `Microsoft.Owin.Security.Jwt` package.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
Auth0 allows you to add authorization to any kind of application. This guide demonstrates how to integrate Auth0 with any new or existing ASP.NET Owin Web API application using the `Microsoft.Owin.Security.Jwt` package.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+If you haven't created an API in your Auth0 dashboard yet, you can use the interactive selector to create a new Auth0 API or select an existing API that represents the project you want to integrate with.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
If you have not created an API in your Auth0 dashboard yet, you can use the interactive selector to create a new Auth0 API or select an existing API for your project. 
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+  - files/scope-authorize-attribute
+  - files/api-controller
+---
+
+# Add authorization to an ASP.NET Owin Web API application.
+Auth0 allows you to quickly add authorization to any kind of application. This guide demonstrates how to integrate Auth0 with any new or existing ASP.NET Owin Web API application using the `Microsoft.Owin.Security.Jwt` package.
+
+If you haven't created an API in your Auth0 dashboard yet, you can use the interactive selector to create a new Auth0 API or select an existing API that represents the project you want to integrate with. 
+
+Alternatively, you can read [our getting started guide](get-started/auth0-overview/set-up-apis) that helps you set up your first API through the Auth0 dashboard.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
To set up your first API through the Auth0 dashboard, review [our getting started guide](get-started/auth0-overview/set-up-apis).
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+
+Every API in Auth0 is configured using an API Identifier that your application code will use as the Audience to validate the access token.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
Each Auth0 API uses the API Identifier, which your application needs to validate the access token.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+<%= include('../../../_includes/_api_auth_intro') %>
+
+## Define permissions
+<%= include('../_includes/_api_scopes_access_resources') %>
+
+
+## Install dependencies
+
+To use Auth0 access tokens with ASP.NET Owin Web API you will use the OWIN JWT Middleware which is available in the `Microsoft.Owin.Security.Jwt` NuGet package.
+
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
Install the `Microsoft.Owin.Security.Jwt` NuGetPackage. This package contains the OWIN JWT Middleware necessary to use Auth0 access tokens in the ASP.NET Owin Web API.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+```
+
+## Configure the middleware {{{ data-action=code data-code="Startup.cs" }}}
+
+Go to the `Configuration` method of your `Startup` class and add a call to `UseJwtBearerAuthentication` passing in the configured `JwtBearerAuthenticationOptions`.
+
+The `JwtBearerAuthenticationOptions` needs to specify your Auth0 API Identifier in the `ValidAudience` property, and the full path to your Auth0 domain as the `ValidIssuer`. You will need to configure the `IssuerSigningKeyResolver` to use the instance of `OpenIdConnectSigningKeyResolver` to resolve the signing key:
+
+::: panel-warning Do not forget the trailing backslash
+Please ensure that the URL specified for `ValidIssuer` contains a trailing backslash as this needs to match exactly with the issuer claim of the JWT. This is a common misconfiguration error which will cause your API calls to not be authenticated correctly.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
Ensure the URL specified for `ValidIssuer` contains a trailing backslash. This must match exactly with the JWT issuer claim. API calls will not authenticate correctly if you misconfigured this value.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 14]
```diff
@@ -0,0 +1,100 @@
...
+:::
+
+## Verifying the token signature {{{ data-action=code data-code="OpenIdConnectSigningKeyResolver.cs" }}}
+As the OWIN JWT middleware doesn't use Open ID Connect Discovery by default, you will need to provide a custom `IssuerSigningKeyResolver`. To do this, create the `OpenIdConnectSigningKeyResolver` class and ensure to return the correct `SecurityKey` by implementing `GetSigningKey`.
+
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
The OWIN JWT middleware does not use Open ID Connect Discovery by default, so you must provide a custom `IssuerSigningKeyResolver`. Create the `OpenIdConnectSigningKeyResolver` class and ensure to return the correct `SecurityKey` by implementing `GetSigningKey`.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 15]
```diff
@@ -0,0 +1,100 @@
...
+This class is then used as `TokenValidationParameters.IssuerSigningKeyResolver` while configuring the middleware in `Startup.cs`.
+
+:::note
+Such a custom resolver was previously published as part of the `Auth0.OpenIdConnectSigningKeyResolver` package through Nuget. As [this package is not available anymore](https://github.com/auth0/auth0-aspnet-owin/blob/master/SECURITY-NOTICE.md), you will need to provide this yourself.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
This custom resolver is deprecated and [no longer available](https://github.com/auth0/auth0-aspnet-owin/blob/master/SECURITY-NOTICE.md). You must provider this customer resolver yourself.
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 16]
```diff
@@ -0,0 +1,100 @@
...
+
+Now that you have configured your application, run your application to verify that:
+* `GET /api/public` is available for non-authenticated requests.
+* `GET /api/private` is available for authenticated requests.
+* `GET /api/private-scoped` is available for authenticated requests containing an access token with the `read:messages` scope.
+
+:::
+
+:::checkpoint-failure
+Sorry about that. Here's a couple things to double check:
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
If your application did not start successfully:
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 17]
```diff
@@ -0,0 +1,100 @@
...
+* make sure `ValidIssuer` and `ValidAudience` are configured correctly
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
* Ensure your configured the  `ValidIssuer` and `ValidAudience` values correctly
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 18]
```diff
@@ -0,0 +1,100 @@
...
+* make sure the token is added as the `Authorization` header
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
* Verify you added the token as the `Authorization` header
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+* does the token have the correct scopes? You can use [jwt.io](https://jwt.io/) to verify.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
* Ensure the token has the correct scopes. Verify with [jwt.io](https://jwt.io/).
```

File: articles/quickstart/backend/webapi-owin/interactive.md [hunk 20]
```diff
@@ -0,0 +1,100 @@
...
+
+Create a class called `ScopeAuthorizeAttribute` which inherits from `System.Web.Http.AuthorizeAttribute`. This attribute will check that the `scope` claim issued by your Auth0 tenant is present, and if so it will ensure that the `scope` claim contains the requested scope.
+
+## Protect API endpoints {{{ data-action=code data-code="ApiController.cs" }}}
+
+<%= include('../_includes/_api_endpoints') %>
+
+The JWT middleware integrates with the standard ASP.NET authentication and authorization mechanisms, so you only need to decorate your controller action with the `[Authorize]` attribute to secure an endpoint.
+
+To ensure that a scope is present in order to call a particular API endpoint, you simply need to decorate the action with the `ScopeAuthorize` attribute and pass the name of the required `scope` in the `scope` parameter.
```
Comment (tr0njavolta) [inline articles/quickstart/backend/webapi-owin/interactive.md]:
```suggestion
Update the action with the `ScopeAuthorize` attribute and pass the name of the required `scope` in the `scope` parameter. This ensures the correct scope is available to call a specific API endpoing.
```

**Review by tr0njavolta — COMMENTED** [review]
Some updates & suggestions

**Review by tr0njavolta — APPROVED** [review]

**Pull request merged**